poc/QtImageRegistration/engine.cpp

```diff
@@ -0,0 +1,158 @@
+
+
+
+    if (image1.empty() || image2.empty()) {
+        QMessageBox::warning(nullptr, "Error", "Error in reading images");
+        return;
+    }
+
+    this->m_referenceImage = image1;
+    this->m_targetImage = image2;
+}
+
+Engine::Engine(QObject *parent, const Mat &referenceImage, const Mat &targetImage) : QObject(parent), m_referenceImage(referenceImage), m_targetImage(targetImage)
+{
+
+}
+
+Mat Engine::align(AlignmentAlgorithms algorithm)
+{
+    Mat alignedImage;
+    switch (algorithm) {
+    case AlignmentAlgorithms::FEATURE_BASED:
+        alignedImage = align_featureBased();
+    case AlignmentAlgorithms::INTENSITY_BASED:
+        alignedImage = align_intensityBased();
+    default:
+        break;
+    }
+
+    return alignedImage;
+}
+
+Mat Engine::align_featureBased()
+{
+    // Convert images to grayscale
+    Mat grayReference, grayTarget;
+    cvtColor(this->m_referenceImage, grayReference, COLOR_BGR2GRAY);
+    cvtColor(this->m_targetImage, grayTarget, COLOR_BGR2GRAY);
+
+    // Initialize SIFT detector and descriptor
+    Ptr<SIFT> sift = SIFT::create();
+
+    // Detect keypoints and compute descriptors
+    vector<KeyPoint> keypointsReference, keypointsTarget;
+    Mat descriptorsReference, descriptorsTarget;
+    sift->detectAndCompute(grayReference, Mat(), keypointsReference, descriptorsReference);
+    sift->detectAndCompute(grayTarget, Mat(), keypointsTarget, descriptorsTarget);
+
+    // Match descriptors
+    BFMatcher matcher(NORM_L2);
+    vector<vector<DMatch>> knnMatches;
+    matcher.knnMatch(descriptorsReference, descriptorsTarget, knnMatches, 2); // k=2 for 2-nearest neighbors
+
+    // Apply ratio test to filter matches
+    vector<DMatch> goodMatches;
+    for (size_t i = 0; i < knnMatches.size(); i++) {
+        if (knnMatches[i][0].distance < 0.75 * knnMatches[i][1].distance) {
+            goodMatches.push_back(knnMatches[i][0]);
+        }
+    }
+
+    // Extract matched keypoints
+    vector<Point2f> matchedPointsReference, matchedPointsTarget;
+    for (size_t i = 0; i < goodMatches.size(); i++) {
+        matchedPointsReference.push_back(keypointsReference[goodMatches[i].queryIdx].pt);
+        matchedPointsTarget.push_back(keypointsTarget[goodMatches[i].trainIdx].pt);
+    }
+
+    // Compute weights for each match (optional, based on distance)
+    vector<float> weights;
+    for (size_t i = 0; i < goodMatches.size(); i++) {
+        float dist = goodMatches[i].distance;
+        weights.push_back(1.0 / (1.0 + dist)); // Weight = 1 / (1 + distance)
+    }
+
+    // Find homography matrix with weighted points
+    Mat homography = findHomography(matchedPointsTarget, matchedPointsReference, RANSAC, 3.0, weights);
+
+    // Warp target image using homography
+    Mat alignedImage;
+    warpPerspective(this->m_targetImage, alignedImage, homography, this->m_referenceImage.size());
+
+    // return homography for the api
+    return alignedImage;
```

Comment:
Are we returning the matrix or the image? The comment says matrix, but the variable name says image. The consumer wants the matrix.

Reply:
yes these 2 last lines are remaining to test previewing result , and have to get commented out when using the API

Reply:
Then please change the comment to "Will change to return homography for the API".

Reply:
Sure, Done